elastic/kibana
# [Alerting] Limit the executable actions per rule execution

fixes: #124871

## Summary

This PR intends to limit the executable actions per Rule execution with a circuit breaker by reading the maxExecutableActions from kibana.yml.
- If a rule execution exceeds the max. number, circuit breaker kicks in and stops triggering actions.
- Sets the rule status `warning` as well as a `reason` and a `message` in the Rule SO.
- Logs an `execute` event with a reason and message (not a separate warning object in the event log, uses the existing reason and message fields)
- Rule status in rules list on the UI becomes 'warning'
- Displays a warning banner on the Rule Details page

### To Verify:
Create a Rule that has multiple actions.
Set the maxExecutableActions in kibana.yml as shown below, lesser than the number of actions you have created
```
xpack.alerting.rules.execution:
  actions:
    max: 2
```
Rule status in rules list on the UI must be `warning` with a respective color (yellow) 
A warning banner is displayed in rule details page (Warning theme will be discussed with the UX team)
numberOfTriggered actions in the event log must be equal to the number you have set in kibana.yml (maxExecutableActions)

Do the same validation with a rule type specific config in kibana.yml as shown below
```
xpack.alerting.rules.execution:
  actions:
    max: 2
  ruleTypeOverrides:
    - id: <your-rule-type-id>
         actions:
           max: 2
```

<img width="1040" alt="list" src="https://user-images.githubusercontent.com/92688503/156763606-84493bf9-6f54-4894-9513-e7df5433dc78.png">

<img width="1026" alt="details" src="https://user-images.githubusercontent.com/92688503/156763620-88a63a4b-9fed-4b54-87d1-542cc08923d9.png">


### Checklist

Delete any items that are not applicable to this PR.
